// react-native-selectmultiple-button/libraries/SelectMultipleButton.js
/*
 * @Author: Young
 * DSHARP
 * @flow 
 * @Date: 2018-02-06 13:54:25 
 * @Last Modified by: Young
 * @Last Modified time: 2018-08-31 14:03:36
 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  View,
  Text,
  TouchableWithoutFeedback,
  Alert,
  StyleSheet,
} from 'react-native'

const ios_blue = '#007AFF'

export default class SelectMultipleButton extends Component {

  static propTypes = {
    selected: PropTypes.bool,

    value: PropTypes.oneOfType(
      [
        PropTypes.string,
        PropTypes.number
      ]
    ).isRequired,
    displayValue: PropTypes.oneOfType(
      [
        PropTypes.string,
        PropTypes.number
      ]
    ),

    highLightStyle: PropTypes.shape({
      borderColor: PropTypes.string.isRequired,
      backgroundColor: PropTypes.string.isRequired,
      textColor: PropTypes.string.isRequired,
      borderTintColor: PropTypes.string.isRequired,
      backgroundTintColor: PropTypes.string.isRequired,
      textTintColor: PropTypes.string.isRequired,
    }),

    buttonViewStyle: PropTypes.object,
    textStyle: PropTypes.object,
    singleTap: PropTypes.func,
	renderLabel: PropTypes.func

  }

  static defaultProps = {
    selected: false,
    highLightStyle: {
      borderColor: 'gray',
      backgroundColor: 'transparent',
      textColor: 'gray',
      borderTintColor: ios_blue,
      backgroundTintColor: 'transparent',
      textTintColor: ios_blue,
	  renderLabel: null
    },

    singleTap: (valueTap) => { },
  }

  constructor(props) {
    super(props)
    this.state = {
      selected: false,
    }
  }

  componentDidMount() {
    this.setState({
      selected: this.props.selected,
    })
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.selected !== nextProps.selected) {
      this.setState({
        selected: nextProps.selected
      })
    }
  }
  
	renderLabel = (label, style, selected) => {
    if (this.props.renderLabel) {
      return this.props.renderLabel(label, style, selected)
    }
    return (
      <Text style={style}>{label}</Text>
    )
  }
  renderLabel = (label, style, selected) => {
    if (this.props.renderLabel) {
      return this.props.renderLabel(label, this.props.textStyle, selected,this.props.highLightStyle)
    }
    return (
      <Text style={style,{ color: this.state.selected ? this.props.highLightStyle.textTintColor : this.props.highLightStyle.textColor }}>{label}</Text>
    )
  }
  render() {
    return (
      <TouchableWithoutFeedback
        onPress={() => {
          this.props.singleTap(this.props.value)
        }
        }>
		{this.renderLabel(this.props.displayValue === undefined ? this.props.value : this.props.displayValue, this.props.textStyle, this.state.selected)}
      </TouchableWithoutFeedback>
    )
  }
}

const styles = StyleSheet.create({
  button: {
    margin: 5,
    borderRadius: 3,
    alignItems: 'center',
    justifyContent: 'center',
    borderWidth: 1
  },
  text: {
    textAlign: 'center',
    marginTop: 5,
    marginBottom: 5,
    marginLeft: 10,
    marginRight: 10,
  }
})
